debug_module.tex

```diff
@@ -18,12 +18,10 @@ \chapter{Debug Module (DM)} \label{dm}
 \item Allow direct System Bus Access. (Optional)
 \end{steps}
 
-In order to implement memory access, a target must implement at least one of
-Program Buffer, System Bus Access, or Abstract Access Memory command
-mechanisms.
-
-In order to implement access to all registers, targets should implement either
-the Program Buffer or abstract access to all registers.
+In order to be compliant with this specification an implementation must
+implement all the required features listed above, at least one of Program Buffer,
+System Bus Access, or Abstract Access Memory command mechanisms, and either the
+Program Buffer or abstract access to all registers.
```

Comment:
"all registers" means "all CSRs and GPRs"? What about FPRs? What about vector registers? I think this is a bit vague. And can't an implementation be "fully compliant" and generally useful for debugging even if e.g. you can't access the FPRs?

Reply:
Let's tackle the second question first:
> can't an implementation be "fully compliant" and generally useful for debugging even if e.g. you can't access the FPRs?

It's better than not having any debugging, but the same could be said for single step, halt out of reset, and a host of other features which are not hard to implement, and provide a lot of value.

A debug solution that can't access all the registers is incomplete at best. If you have a vector unit, your debug implementation must provide access to its registers. Use some of the non-standard extension space in abstract commands if you have to.

In what kind of system would you choose not to implement debug access to all registers?

Reply:
I'm just trying to make it clear, I think "registers" is a little fuzzy of a term.

Reply:
I've pushed a change that hopefully is more clear.